### PR TITLE
!!! TASK: Interdimensional relatives for node move

### PR DESCRIPTION
**Pull request by nezaniel**
Builds on: #4982 

This replaces the cumbersome MoveNodeMappings DTO with the far more lightweight, but surprisingly even more capable InterdimensionalSiblings DTO.

**Upgrade instructions**

This is breaky for 3rd party packages since the event DTOs are part of the API

**Review instructions**

First: Don't be afraid of the number of LoC added. That's just proper test coverage for MoveNode, finally.

For the refactoring, further assumptions are made (and enforced via constraint checks and tests):

1) Pure edge operation
MoveNode is considered an pure edge operation, meaning no node is actually touched. This was also implied before, since MoveNode explicitly never updated any timestamps. Thus, no OrginDimensionSpacePoints are communicated anymore, just siblings per affected DSP.
Hint: Neos' ChangeProjection needs to decide how to deal with MoveNode. Its tests pass anyway.

2) Only one parent (at most) per move operation
As the RelationDistributionStrategy already states: the variants will be gathered (more or less, depending on the strategy) at the new parent, if given.
That also means: no implicit parent changes. If a given sibling belongs to a different parent, this will lead to
* a constraint check exception, if it's not really a sibling in the selected DSP
* the "sibling" being ignored in an affected DSP if it's not a sibling there

3) Succeeding sibling has priority over preceding sibling
The given succeeding sibling (and its further succeeding siblings) will be checked for applicability.
If none is found and a preceding sibling is given, it (and its further preceding siblings) will be checked instead.

4) Preceding siblings are evaluated if given and no succeeding sibling is given

5) Non-determinable siblings
If no succeeding sibling can be determined for the move operation, there are three cases
* both siblings are explicitly given as null, which reflects the intention of "move to end"
* a parent id is given; We need to define a succeeding sibling (or null) here and chose null
* no parent is given: This is a no-op in the affected DSP; It will not even be part of the InterdimensionalSiblings because nothing happens in this DSP

**Checklist**

- [x] Code follows the PSR-12 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the 9.0 branch
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

File: Neos.ContentRepository.Core/Classes/Projection/ContentGraph/NodeTags.php [hunk 1]
```diff
@@ -33,8 +33,8 @@
 final readonly class NodeTags implements \IteratorAggregate, \Countable, \JsonSerializable
 {
     private function __construct(
-        private SubtreeTags $tags,
-        private SubtreeTags $inheritedTags,
+        public SubtreeTags $tags,
+        public SubtreeTags $inheritedTags,
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Projection/ContentGraph/NodeTags.php]: How is that related? I deliberately left those private to simplify the API

File: Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/ProjectedNodeTrait.php [hunk 2]
```diff
@@ -650,6 +650,28 @@ public function iExpectThisNodeToHaveTheFollowingSucceedingSiblings(TableNode $e
         });
     }
 
+    /**
+     * @Then /^I expect this node to tag with "(.*)"$/
```
Comment (bwaidelich) [inline Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/ProjectedNodeTrait.php]: "expect the node to tag with" does not really make sense to me. We should find a common nomenclature.
Also: There were already similar behat steps that didn't require API changes IIRC

File: Neos.Neos/Classes/PendingChangesProjection/ChangeProjection.php [hunk 3]
```diff
@@ -230,18 +231,22 @@ private function whenRootWorkspaceWasCreated(RootWorkspaceWasCreated $event): vo
...
-        $this->markAsMoved(
-            $event->getContentStreamId(),
-            $event->getNodeAggregateId(),
-            $mapping[0]->movedNodeOrigin
-        );
+        // Changes reflect the editorial intention, not the effect as they are later published via commands
+        $metadata = $eventEnvelope->event->metadata?->value ?? [];
+        if (isset($metadata['commandPayload'])) {
+            $command = MoveNodeAggregate::fromArray($metadata['commandPayload']);
+            // WORKAROUND: we simply use the command's DSP here as the origin dimension space point.
```
Comment (bwaidelich) [inline Neos.Neos/Classes/PendingChangesProjection/ChangeProjection.php]: As far as I know we only relied on the command-metadata in the write side for rebasing, right?
I still hope that we can replace that at some point with a pure event based approach.
I don't fully understand the change yet, but I would really try to keep that implementation detail as local as possible and it makes me think: Do we maybe miss properties in the event if we have to re-create the command at projection time?

**Review by bwaidelich — COMMENTED** [review]
Impressive, thanks for taking care!
Just a first batch of comments after skimming over from my phone

File: Neos.Neos/Classes/PendingChangesProjection/ChangeProjection.php (hunk 3 above)
Comment (nezaniel) [reply Neos.Neos/Classes/PendingChangesProjection/ChangeProjection.php]: I think the issue here is with the change projection. If it needs to know about origin DSPs it has to keep track of them itself, then the event will suffice again. Then again, I don't really know what the change projection does with MoveNode anyway. This is to be discussed elsewhere and I am pretty sure this workaround can then be removed.

File: Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/ProjectedNodeTrait.php (hunk 2 above)
Comment (nezaniel) [reply Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/ProjectedNodeTrait.php]: Agreed, I'm not happy with that either. 
The behat steps that I found only let me check for a single tag, as did the API. But here I need to check whether this node uses the given tags, all of them and only them.

File: Neos.ContentRepository.Core/Classes/Projection/ContentGraph/NodeTags.php (hunk 1 above)
Comment (nezaniel) [reply Neos.ContentRepository.Core/Classes/Projection/ContentGraph/NodeTags.php]: I need to check the complete set of tags against a list, see below

File: Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/ProjectedNodeTrait.php (hunk 2 above)
Comment (kitsunet) [reply Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/ProjectedNodeTrait.php]: "I expect this node is tagged with ..." ?

File: Neos.Neos/Classes/PendingChangesProjection/ChangeProjection.php (hunk 3 above)
Comment (kitsunet) [reply Neos.Neos/Classes/PendingChangesProjection/ChangeProjection.php]: this is an interesting question, for whom do we make the events, something we should _really_ define for the future as well. As in do we record only whatever at that point is relevant to the content graph or anything that might be of interested at this particular point?

File: Neos.ContentRepository.Core/Classes/Projection/ContentGraph/NodeTags.php (hunk 1 above)
Comment (bwaidelich) [reply Neos.ContentRepository.Core/Classes/Projection/ContentGraph/NodeTags.php]: I don't think that this is required, since there are `withoutInherited()` and `onlyInherited()` convenience methods.
It's not a big difference, but my goal was to make the API of `SubtreeTags` and `NodeTags` as similar as possible because in most cases you are only interested whether a node has a certain tag or not and iteration etc should always be the same.

File: Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/ProjectedNodeTrait.php (hunk 2 above)
Comment (bwaidelich) [reply Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/ProjectedNodeTrait.php]: I went for
```
I expect the node with aggregate identifier "<foo>" to be explicitly tagged "<bar>"
```
and
```
I expect the node with aggregate identifier "<foo>" to inherit the tag "<bar>"
```

in the [ProjectedNodeTrait](https://github.com/neos/neos-development-collection/blob/7977bac7dc7fb8e4bf1f7accb5668d5e005fa69b/Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/ProjectedNodeTrait.php#L199-L227) and I think we could just re-use those step implementations!?

File: Neos.Neos/Classes/PendingChangesProjection/ChangeProjection.php (hunk 3 above)
Comment (bwaidelich) [reply Neos.Neos/Classes/PendingChangesProjection/ChangeProjection.php]: I need to understand this case better – but in general I would say that we consider the "metadata" of events internal

File: Neos.ContentRepository.Core/Classes/Projection/ContentGraph/NodeTags.php (hunk 1 above)
Comment (nezaniel) [reply Neos.ContentRepository.Core/Classes/Projection/ContentGraph/NodeTags.php]: Ah, you're right, without/onlyInherited()->toStringArray does the job just as good. I reverted this.

File: Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/ProjectedNodeTrait.php (hunk 2 above)
Comment (nezaniel) [reply Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/ProjectedNodeTrait.php]: I want to do a more exact comparison, but I adjusted the naming to be more in line

File: Neos.Neos/Classes/PendingChangesProjection/ChangeProjection.php (hunk 3 above)
Comment (nezaniel) [reply Neos.Neos/Classes/PendingChangesProjection/ChangeProjection.php]: We don't make the events for anyone, we just need to communicate the effect of what happened. And I'm pretty sure that's what we do ^^
If reading the metadata is the issue we can as well use the first affected DSP instead of the explicitly selected one, should be the same.

Comment (nezaniel) [reply Neos.Neos/Classes/PendingChangesProjection/ChangeProjection.php]: ...and adjusted.

File: Neos.ContentGraph.DoctrineDbalAdapter/Tests/Behavior/Features/Projection/SiblingPositions.feature [hunk 4]
```diff
@@ -0,0 +1,110 @@
+@contentrepository
+Feature: Sibling positions are properly resolved
+
+  In the general DBAL adapter, hierarchy relations are sorted by an integer field. It defaults to a distance of 128,
+  which is reduced each time a node is inserted between two siblings. Once the number becomes uneven, the siblings positions are recalculated.
+  These are the test cases for this behavior.
```
Comment (bwaidelich) [inline Neos.ContentGraph.DoctrineDbalAdapter/Tests/Behavior/Features/Projection/SiblingPositions.feature:6]: perfect summary, that really helps!

File: Neos.ContentRepository.Core/Classes/Feature/NodeMove/Event/NodeAggregateWasMoved.php [hunk 5]
```diff
@@ -95,16 +76,43 @@ public function createCopyForContentStream(ContentStreamId $targetContentStreamI
...
+                    );
+                }
+            }
+            $succeedingSiblingsForCoverage = new InterdimensionalSiblings(...$succeedingSiblings);
+        } else {
+            $newParentNodeAggregateId = $values['newParentNodeAggregateId'] === null
+                ? null
+                : NodeAggregateId::fromString($values['newParentNodeAggregateId']);
+            $succeedingSiblingsForCoverage = InterdimensionalSiblings::fromArray($values['succeedingSiblingsForCoverage']);
+        }
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/NodeMove/Event/NodeAggregateWasMoved.php:109]: This is the "migration" right?
In other places we (or rather @mhsdesign) decided to go for an event migration instead.

I'm not really decided on that matter but my tendency is to go for an event migration until we have tagged a first RC
In any case we should probably agree on a common strategy for these cases

File: Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Nodes.php [hunk 6]
```diff
@@ -194,4 +197,12 @@ public function nextAll(Node $referenceNode): self
...
         return new self(array_slice($this->nodes, $referenceNodeIndex + 1));
     }
+
+    public function getNodeAggregateIds(): NodeAggregateIds
+    {
+        return NodeAggregateIds::create(...array_map(
+            fn (Node $node): NodeAggregateId => $node->nodeAggregateId,
+            $this->nodes
+        ));
+    }
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Nodes.php]: Just a nitpick, but I tend to add more generic `map` and (if needed) `filter` methods to my collection types like:
```suggestion
    /**
      * @param \Closure(Node $node): mixed $callback
      * @return array<mixed>
      **/
    public function map(\Closure $callback): NodeAggregateIds
    {
        return array_map($callback, $this->nodes);
        return NodeAggregateIds::create(...array_map(
            fn (Node $node): NodeAggregateId => $node->nodeAggregateId,
            $this->nodes
        ));
    }
```

With that, we could write:

```php
NodeAggregateIds::fromArray($nodes->map(fn (Node $n) => $node-> nodeAggregateId));
```

but use it for other mappings as well.
We could still keep `getNodeAggregateIds()` for convenience..

Anyways, just a mini nitpick ofc

**Review by bwaidelich — APPROVED** [review]
+1 by reading.
Thanks a lot for catching me up earlier, I finally understood it and it makes a lot of sense :)
I left some comments re migration strategy, but no reason to block this IMO.

Also, what are those:
image

?

File: Neos.ContentRepository.Core/Classes/Feature/NodeMove/Event/NodeAggregateWasMoved.php (hunk 5 above)
Comment (nezaniel) [reply Neos.ContentRepository.Core/Classes/Feature/NodeMove/Event/NodeAggregateWasMoved.php:109]: I'd be fine with a migration as well, I just thought we wouldn't do these anymore ^^

**Comment by nezaniel** [comment]
> Also, what are those:
> [...]
> ?

It's untouched; I guess it fails on a newer / older PHPStan version

File: Neos.ContentRepository.Core/Classes/Feature/NodeMove/Event/NodeAggregateWasMoved.php (hunk 5 above)
Comment (nezaniel) [reply Neos.ContentRepository.Core/Classes/Feature/NodeMove/Event/NodeAggregateWasMoved.php:109]: I'd say we write event migrations for InterdimensionalSiblings in a separate PR, both for node variation and move

**Comment by nezaniel** [comment]
...and was fixed by just merging 9.0

File: Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/Feature/NodeMove.php [hunk 7]
```diff
@@ -39,166 +36,154 @@ abstract protected function getTableNamePrefix(): string;
...
-            $succeedingSiblingNodeMoveDestination->parentOriginDimensionSpacePoint,
-        );
-        if ($newParent === null) {
-            // this should NEVER happen; because this would mean $newSucceedingSibling
-            // wouldn't have a parent => invariant violation.
-            throw EventCouldNotBeAppliedToContentGraph::becauseTheTargetSucceedingSiblingNodesParentIsMissing(
-                MoveNodeAggregate::class
+        $newSucceedingSibling = null;
+        if ($succeedingSiblingForCoverage->nodeAggregateId) {
+            // find the new succeeding sibling NodeRecord; We need this records because we'll use its RelationAnchorPoint later.
```
Comment (grebaldi) [inline Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/Feature/NodeMove.php]:
```suggestion
            // find the new succeeding sibling NodeRecord; We need this record because we'll use its RelationAnchorPoint later.
```

File: Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/Feature/NodeMove.php [hunk 8]
```diff
@@ -39,166 +36,154 @@ abstract protected function getTableNamePrefix(): string;
...
      * This helper is responsible for moving a single incoming HierarchyRelation of $nodeToBeMoved
-     * to a new location. $coveredDimensionSpacePointWhereMoveShouldHappen specifies which incoming HierarchyRelation
-     * should be moved.
+     * to a new location including a change of parent. $succeedingSiblingForCoverage specifies
+     * which incoming HierarchyRelation should be moved and where exactly.
      *
-     * The move target is given as $parentNodeMoveTarget. We always move to the END of the children list of the
-     * given parent.
+     * The move target is given as $parentNodeAggregateId and $succeedingSiblingForCoverage.
+     * We always move to parent after the succeeding sibling if given (or to the end)
```
Comment (grebaldi) [inline Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/Feature/NodeMove.php]: This confuses me. Specifically

> We always move **to** parent **after** the **succeeding** sibling

I take that `to` refers to `beneath the parent`. But I can't wrap my head around how we move anything "after" the "succeeding" sibling. Or am I misinterpreting what the succeeding sibling is? Up until now, I have understood the `$succeedingSiblingForCoverage` to be the future succeeding sibling of the node that is being moved. But if that is true, then we are hardly moving the node *after* that sibling (because it would be the preceding sibling then), right?

Can you shed some light on which part of the code comment and/or my interpretation of things is correct/incorrect, @nezaniel?

File: Neos.Neos/Classes/PendingChangesProjection/ChangeProjection.php [hunk 9]
```diff
@@ -183,7 +184,7 @@ public function apply(EventInterface $event, EventEnvelope $eventEnvelope): void
     {
         match ($event::class) {
             RootWorkspaceWasCreated::class => $this->whenRootWorkspaceWasCreated($event),
-            NodeAggregateWasMoved::class => $this->whenNodeAggregateWasMoved($event),
+            NodeAggregateWasMoved::class => $this->whenNodeAggregateWasMoved($event, $eventEnvelope),
```
Comment (grebaldi) [inline Neos.Neos/Classes/PendingChangesProjection/ChangeProjection.php]:
```suggestion
            NodeAggregateWasMoved::class => $this->whenNodeAggregateWasMoved($event),
```

`$eventEnvelope` does not appear to be used.

File: Neos.Neos/Classes/PendingChangesProjection/ChangeProjection.php [hunk 10]
```diff
@@ -230,18 +231,23 @@ private function whenRootWorkspaceWasCreated(RootWorkspaceWasCreated $event): vo
         }
     }
 
-    private function whenNodeAggregateWasMoved(NodeAggregateWasMoved $event): void
+    private function whenNodeAggregateWasMoved(NodeAggregateWasMoved $event, EventEnvelope $eventEnvelope): void
```
Comment (grebaldi) [inline Neos.Neos/Classes/PendingChangesProjection/ChangeProjection.php]:
```suggestion
    private function whenNodeAggregateWasMoved(NodeAggregateWasMoved $event): void
```

File: Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Nodes.php [hunk 11]
```diff
@@ -194,4 +197,20 @@ public function nextAll(Node $referenceNode): self
...
+    /**
+     * @param \Closure(Node $node): mixed $callback
+     * @return array<mixed>
+     */
+    public function map(\Closure $callback): array
+    {
+        return array_map($callback, $this->nodes);
+    }
+
+    public function mapToNodeAggregateIds(): NodeAggregateIds
```
Comment (grebaldi) [inline Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Nodes.php]:
```suggestion
    public function toNodeAggregateIds(): NodeAggregateIds
```

otherwise confusing, imho

**Review by grebaldi — COMMENTED** [review]
Hi @nezaniel,

I went through your code, and I believe I understand the gist of it :sweat_smile: Left some comments, none of which are big things.

I also did some manual testing with the UI and ran the UI E2E tests atop your branch, and everything looks :+1:

File: Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/Feature/NodeMove.php (hunk 8 above)
Comment (nezaniel) [reply Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/Feature/NodeMove.php]: Err, you're right of course. We move stuff *beneath* the parent and *before* the succeeding sibling or to the end.

File: Neos.Neos/Classes/PendingChangesProjection/ChangeProjection.php (hunk 9 above)
Comment (nezaniel) [reply Neos.Neos/Classes/PendingChangesProjection/ChangeProjection.php]: Yep, it used to, but no more

File: Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Nodes.php (hunk 11 above)
Comment (nezaniel) [reply Neos.ContentRepository.Core/Classes/Projection/ContentGraph/Nodes.php]: Fine by me

File: Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/Feature/NodeMove.php (hunk 8 above)
Comment (nezaniel) [reply Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/Feature/NodeMove.php]: here you go

Comment (grebaldi) [reply Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/Feature/NodeMove.php]: Thanks :)

**Review by grebaldi — APPROVED** [review]

File: Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/ProjectedNodeTrait.php [hunk 12]
```diff
@@ -241,6 +241,28 @@ public function iExpectTheNodeWithAggregateIdentifierToNotContainTheTag(string $
...
     }
 
+    /**
+     * @Then /^I expect this node to be exactly explicitly tagged "(.*)"$/
+     * @param string $tagList the comma-separated list of tag names
+     */
+    public function iExpectThisNodeToBeExactlyExplicitlyTagged(string $tagList): void
+    {
+        $this->assertOnCurrentNode(function (Node $currentNode) use ($tagList) {
+            $currentNode->tags->withoutInherited()->toStringArray() === explode(',', $tagList);
```
Comment (mhsdesign) [inline Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/ProjectedNodeTrait.php:251]: lol i just noticed this is a false positive test. There are actually no assertions here but just an expression and we return void.

`Assert::assertEquals` must be used for this test to have an effect i guess?

Or am i missing something?

Comment (bwaidelich) [reply Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/ProjectedNodeTrait.php:251]: Good catch! o.O

Comment (nezaniel) [reply Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/ProjectedNodeTrait.php:251]: Thanks a lot!

This seems to be a real false positive, since there seem to be issues with MoveNode vs. Subtree tags. I'm currently investigating

Comment (nezaniel) [reply Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/ProjectedNodeTrait.php:251]: The fix for this is applied here:
https://github.com/neos/neos-development-collection/pull/5025
